### ISSUE TITLE
Stuck in compilation (Theano backend)

### ISSUE THREAD

**Issue by mikekestemont** [fchollet/keras#1607]
I am experiencing compilation issues using the Theano backend for this model:
[https://github.com/mikekestemont/pandora/blob/master/pandora/model.py](https://github.com/mikekestemont/pandora/blob/master/pandora/model.py)

The problem seems to lie with the recurrent LSTM encoder on lines 97-139. With bleeding edge installations of Keras and Theano, the model runs fine (and compiles quickly) on the CPU of my local machine, but fails to compile within a reasonable time on the server's GPU (Ubuntu). In fact, Theano seems to get stuck in the compilation phase. A couple of other users reported similar issues in the past days:
#1601 ("The scripts hangs while initializing the net, giving no error nor warning")
#1600 ("Stuck in compiling")
#1520 ("Theano doesn't crash, it hangs, with no error message that could be used to diagnose the problem. This happens during the compilation step.")

This is a vague question, but is there any chance that these errors are in some way related? Thread #1520 suggests a relation to T.dimshuffle? What about the Repeat layer?

**Comment by fchollet** [comment]
Possible. Try it with TensorFlow.

**Comment by farizrahman4u** [comment]
ping @nouiz

**Comment by sunil-at-gh** [comment]
The problem seems to be in `T.stack`. Theano documentation says that the `T.stack(*tensors)` interface used in `K.repeat()` is deprecated. I tried using `T.stack(tensors)` but that also did not seem to work. Hence the proposed change in #1600 .

**Comment by fchollet** [comment]
You can submit a PR with your fix, I'll check it out.

**Comment by nouiz** [comment]
There is theano.tensor.extra_ops.repeat that exist, it is probably what you
want (with axis=0 probably)
http://deeplearning.net/software/theano/library/tensor/extra_ops.html#theano.tensor.extra_ops.repeat

The tile is also a good way to implement this. I'm pretty sure using repeat
or tile is better then using the old implementation, so even if this get
fixed in Theano, I would recommand to change the code to use repeat or tile.

The loop seem strange, I'll try to investigate it.

**Comment by nouiz** [comment]
I have a PR in Theano that make the compilation for all function in that example to only 45s (with a full cache of c code):

https://github.com/Theano/Theano/pull/3974

thanks for the report.

**Comment by nouiz** [comment]
I think we should look for other call in extract_constant or get_scalar_constant_value() in tensor/opt.

Maybe we should change the default of extract_constant to only process constant.

**Comment by sunil-at-gh** [comment]
Sorry, I am not familiar with the process, or that familiar with git.
But looks like it is getting looked at?

_
Sunil

**Comment by nouiz** [comment]
We merged the fix in Theano. So you can just update Theano to the dev version and have the compilation speed up.

**Comment by fchollet** [comment]
In Keras, I have also changed the implementation of `repeat` to use Theano's `repeat` instead of `stack`:

```python
x = x.dimshuffle((0, 'x', 1))
T.extra_ops.repeat(x, n, axis=1)
```

So you can update Keras OR Theano ;-)

**Comment by mikekestemont** [comment]
Awesome, a huge thanks to everyone involved in the fix!

**Comment by sunil-at-gh** [comment]
Thanks! 
_Sunil
